### PR TITLE
PACT-687: Fix issues with date validation in Editorial Web Application

### PR DESCRIPTION
**Pull request by userjd**

### CODE REVIEW

File: app/uk/gov/nationalarchives/omega/editorial/support/DateParser.scala [hunk 1]
```diff
@@ -44,12 +44,35 @@ import scala.util.Try
 object DateParser {
 
   private val dateTimeFormatter =
```
Comment (tna-erasmos) [inline app/uk/gov/nationalarchives/omega/editorial/support/DateParser.scala:46]: Here's a suggested implementation, where we only have to parse once:

``` 
object DateParser {

  private val partDelimiter = "/"

  private val dateTimeFormatter =
    DateTimeFormatter
      .ofPattern(s"d${partDelimiter}M${partDelimiter}y")
      .withLocale(Locale.UK)
      .withResolverStyle(ResolverStyle.SMART)

  def parse(rawDateAsYearMonthDay: String): Option[LocalDate] =
    Try(LocalDate.parse(rawDateAsYearMonthDay, dateTimeFormatter)).toOption
      .filter(dateIsSameAsOriginal(_, rawDateAsYearMonthDay))

  private def dateIsSameAsOriginal(date: LocalDate, rawDateAsYearMonthDay: String): Boolean =
    dateTimeFormatter.format(date) == normaliseFormat(rawDateAsYearMonthDay)

  private def normaliseFormat(originalRawDate: String): String = {
    val partsOfOriginalRawDateWithoutLeadingZeros =
      originalRawDate
        .split(partDelimiter)
        .map(part => String.valueOf(part.toInt))
        .toSeq
    partsOfOriginalRawDateWithoutLeadingZeros.mkString(partDelimiter)
  }

}
```

**Review by tna-erasmos — CHANGES_REQUESTED** [review]
In at least one case, there's a change that wasn't requested.

Also, I noticed in manual testing that if the user entered a leading zero in either day or month, then this would be preserved after saving. This behaviour might be correct, but it wasn't mentioned in the story.

**Comment by tna-erasmos** [comment]
I've added several comments to [the story](https://national-archives.atlassian.net/browse/PACT-687) which you might want to read, @userjd .

File: app/uk/gov/nationalarchives/omega/editorial/support/DateParser.scala (hunk 1 above)
Comment (userjd) [reply app/uk/gov/nationalarchives/omega/editorial/support/DateParser.scala:46]: The above code snippet does't accept zero prefixed date, month and year. In the proposed implementation `LocalDate.of` would allow for this.

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala [hunk 2]
```diff
@@ -71,17 +71,85 @@ class DateParserSpec extends WordSpec with MustMatchers {
...
         Seq("14-9-2020", "9/2020", "2020", "1/14/9/2020")
-          .foreach(rawDate => parse(rawDate) mustBe empty)
+          .foreach(rawDate => parseDate(rawDate) mustBe empty)
       }
       "non existent" in {
         Seq("29/2/2022", "30/2/2022", "31/2/2022", "42/10/2022", "10/14/2022")
-          .foreach(rawDate => parse(rawDate) mustBe empty)
+          .foreach(rawDate => parseDate(rawDate) mustBe empty)
+      }
+      "year is more than 4 digits" in {
```
[the remark this answers is not in the document]
Comment (userjd) [reply test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala:150]: This isn't the requirement. This test is for the validation is to allow upto 4 digits on the year. So the valid values are between 1-9999

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala [hunk 3]
```diff
@@ -23,14 +23,14 @@ package uk.gov.nationalarchives.omega.editorial.support
...
 
 import java.time.LocalDate
 import java.time.Month.{ APRIL, DECEMBER, SEPTEMBER }
 
 class DateParserSpec extends WordSpec with MustMatchers {
 
-  "'parse' returns the expected the result" when {
+  "'parseDate' returns the expected the result" when {
     "the date should be considered valid" when {
       "before the switchover period" in {
```
[the remark this answers is not in the document]
Comment (userjd) [reply test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala:35]: There are tests already validating to allow years from 1 to 9999

File: app/uk/gov/nationalarchives/omega/editorial/support/DateParser.scala (hunk 1 above)
Comment (userjd) [reply app/uk/gov/nationalarchives/omega/editorial/support/DateParser.scala:48]: Done

**Comment by tna-erasmos** [comment]
@userjd , could you format?

**Comment by tna-erasmos** [comment]
Looks good from the outside; here's [my QA video](https://drive.google.com/file/d/1a3gnRJHzh-WsxQqWsPTymODTq0FKxEwm/view?usp=sharing).

I'll review the code, now.

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala:35]: I can see that you've covered this under `"year is 1, 2 or 3 digits"`.

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala [hunk 4]
```diff
@@ -73,8 +73,47 @@ class DateParserSpec extends WordSpec with MustMatchers {
...
+        )((rawDate: String, expectedDate: LocalDate) => parse(rawDate) mustBe Some(expectedDate))
+      }
+
+      "year is 1, 2 or 3 digits" in {
+        forAll(
+          Table(
+            ("Raw Date", "Expected Date"),
+            ("1/12/1", LocalDate.of(1, DECEMBER, 1)),
+            ("5/9/10", LocalDate.of(10, SEPTEMBER, 5)),
+            ("5/9/1000", LocalDate.of(1000, SEPTEMBER, 5))
```
Comment (tna-erasmos) [inline test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala:109]: This is 4 digits.

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala [hunk 5]
```diff
@@ -71,17 +71,85 @@ class DateParserSpec extends WordSpec with MustMatchers {
...
+      "month starting with a zero" in {
+        forAll(
+          Table(
+            ("Raw Date", "Expected Date"),
+            ("29/04/2022", LocalDate.of(2022, APRIL, 29))
+          )
+        )((rawDate: String, expectedDate: LocalDate) => parseDate(rawDate) mustBe Some(expectedDate))
+      }
+
+      "year starting with a zero" in {
```
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala:94]: It's still there.

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala [hunk 6]
```diff
@@ -73,8 +73,47 @@ class DateParserSpec extends WordSpec with MustMatchers {
...
+            ("09/4/2022", LocalDate.of(2022, APRIL, 9))
+          )
+        )((rawDate: String, expectedDate: LocalDate) => parse(rawDate) mustBe Some(expectedDate))
+      }
+
+      "month starting with a zero" in {
+        forAll(
+          Table(
+            ("Raw Date", "Expected Date"),
+            ("29/04/2022", LocalDate.of(2022, APRIL, 29))
```
Comment (tna-erasmos) [inline test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala:89]: If we only have a single param set, we don't really need the table.

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala [hunk 7]
```diff
@@ -73,8 +73,47 @@ class DateParserSpec extends WordSpec with MustMatchers {
           )
         )((rawDate: String, expectedDate: LocalDate) => parse(rawDate) mustBe Some(expectedDate))
       }
+      "day starting with a zero" in {
+        forAll(
+          Table(
+            ("Raw Date", "Expected Date"),
+            ("09/4/2022", LocalDate.of(2022, APRIL, 9))
```
Comment (tna-erasmos) [inline test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala:80]: If we only have a single param set, we don't really need the table.

File: app/uk/gov/nationalarchives/omega/editorial/support/DateParser.scala (hunk 1 above)
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/support/DateParser.scala:48]: Confirmed.

I know that I suggested this, but I'm now thinking that this:

```
.ofPattern("d" + partDelimiter + "M" + partDelimiter + "y")
```

or this:

```
.ofPattern(Seq("d", "M", "y").mkString(partDelimiter))
```

would be clearer than this:

```
.ofPattern(s"d${partDelimiter}M${partDelimiter}y")
```

But I'm OK with any of them, so resolving.

File: app/uk/gov/nationalarchives/omega/editorial/forms/EditSetRecordFormValues.scala [hunk 8]
```diff
@@ -53,12 +53,12 @@ object EditSetRecordFormValues {
       coveringDates = formValues.coveringDates,
       formerReferenceDepartment = formValues.formerReferenceDepartment,
       formerReferencePro = formValues.formerReferencePro,
-      startDateDay = formValues.startDateDay,
-      startDateMonth = formValues.startDateMonth,
-      startDateYear = formValues.startDateYear,
-      endDateDay = formValues.endDateDay,
-      endDateMonth = formValues.endDateMonth,
-      endDateYear = formValues.endDateYear,
+      startDateDay = String.valueOf(formValues.startDateDay.toInt),
```
Comment (tna-erasmos) [inline app/uk/gov/nationalarchives/omega/editorial/forms/EditSetRecordFormValues.scala:56]: If I revert these changes, all of the current tests pass.

Manually, I see that leading zeros are preserved, so I understand the value here.

But perhaps we're missing some scenerios at the controller level (`EditSetRecordControllerSpec`)?

File: app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetController.scala [hunk 9]
```diff
@@ -31,6 +31,7 @@ import uk.gov.nationalarchives.omega.editorial.services.{ EditSetPagination, Edi
 import uk.gov.nationalarchives.omega.editorial.support.{ FormSupport, MessageSupport }
 import uk.gov.nationalarchives.omega.editorial.views.html.editSet
 
+
 import javax.inject._
 
 /** This controller creates an `Action` to handle HTTP requests to the application's home page.
```
Comment (tna-erasmos) [inline app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetController.scala:37]: Was this an accidental change?

**Review by tna-erasmos — CHANGES_REQUESTED** [review]

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala]: Confirmed.

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala [hunk 10]
```diff
@@ -71,17 +71,85 @@ class DateParserSpec extends WordSpec with MustMatchers {
...
+      "year starting with a zero" in {
+        forAll(
+          Table(
+            ("Raw Date", "Expected Date"),
+            ("5/9/022", LocalDate.of(22, SEPTEMBER, 5))
+          )
+        )((rawDate: String, expectedDate: LocalDate) => parseDate(rawDate) mustBe Some(expectedDate))
+      }
+
+      "year is 1 digit" in {
```
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala]: Confirmed.

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala [hunk 11]
```diff
@@ -71,17 +71,85 @@ class DateParserSpec extends WordSpec with MustMatchers {
...
+        forAll(
+          Table(
+            ("Raw Date", "Expected Date"),
+            ("1/12/1", LocalDate.of(1, DECEMBER, 1))
+          )
+        )((rawDate: String, expectedDate: LocalDate) => parseDate(rawDate) mustBe Some(expectedDate))
+      }
+
+      "year is 2 digits" in {
```
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala]: Confirmed.

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala [hunk 12]
```diff
@@ -71,17 +71,85 @@ class DateParserSpec extends WordSpec with MustMatchers {
...
+      "year is 3 digits" in {
+        forAll(
+          Table(
+            ("Raw Date", "Expected Date"),
+            ("5/9/100", LocalDate.of(100, SEPTEMBER, 5))
+          )
+        )((rawDate: String, expectedDate: LocalDate) => parseDate(rawDate) mustBe Some(expectedDate))
+      }
+
+      "year is 4 digits" in {
```
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala]: Confirmed.

File: app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetController.scala (hunk 9 above)
Comment (userjd) [reply app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetController.scala:37]: Yes, I'll revert this

File: app/uk/gov/nationalarchives/omega/editorial/forms/EditSetRecordFormValues.scala (hunk 8 above)
Comment (userjd) [reply app/uk/gov/nationalarchives/omega/editorial/forms/EditSetRecordFormValues.scala:56]: Tests added in `EditSetRecordControllerSpec`

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala:84]: Confirmed.

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala [hunk 13]
```diff
@@ -71,17 +71,85 @@ class DateParserSpec extends WordSpec with MustMatchers {
...
+      "year is 2 digits" in {
+        forAll(
+          Table(
+            ("Raw Date", "Expected Date"),
+            ("5/9/10", LocalDate.of(10, SEPTEMBER, 5))
+          )
+        )((rawDate: String, expectedDate: LocalDate) => parseDate(rawDate) mustBe Some(expectedDate))
+      }
+
+      "year is 3 digits" in {
```
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala]: Confirmed.

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala [hunk 14]
```diff
@@ -71,17 +71,85 @@ class DateParserSpec extends WordSpec with MustMatchers {
             ("20/4/1984", LocalDate.of(1984, APRIL, 20)),
             ("26/12/2022", LocalDate.of(2022, DECEMBER, 26))
           )
-        )((rawDate: String, expectedDate: LocalDate) => parse(rawDate) mustBe Some(expectedDate))
+        )((rawDate: String, expectedDate: LocalDate) => parseDate(rawDate) mustBe Some(expectedDate))
+      }
+      "day starting with a zero" in {
```
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala:76]: Confirmed.

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala (hunk 6 above)
Comment (tna-erasmos) [reply test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala]: Confirmed.

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala (hunk 7 above)
Comment (tna-erasmos) [reply test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala]: Confirmed.

File: app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetController.scala (hunk 9 above)
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/controllers/EditSetController.scala]: Confirmed.

File: app/uk/gov/nationalarchives/omega/editorial/forms/EditSetRecordFormValues.scala (hunk 8 above)
Comment (tna-erasmos) [reply app/uk/gov/nationalarchives/omega/editorial/forms/EditSetRecordFormValues.scala:56]: Good. I like the arrangements.

But I do have a few concerns noted in other comments.

File: test/controllers/EditSetRecordControllerSpec.scala [hunk 15]
```diff
@@ -1272,6 +1272,107 @@ class EditSetRecordControllerSpec extends BaseSpec {
...
+            val editRecordPageResponse = submitWhileLoggedIn("save", 1, oci, values)
+
+            status(editRecordPageResponse) mustBe SEE_OTHER
+            redirectLocation(editRecordPageResponse) mustBe Some(s"/edit-set/1/record/$oci/edit/save")
+
+            val getRecordResult = getRecordForEditingWhileLoggedIn(1, oci)
+            assertPageAsExpected(asDocument(getRecordResult), generateExpectedEditRecordPageFromRecord(oci))
+          }
+        }
+        "end date has a leading zero " when {
```
Comment (tna-erasmos) [inline test/controllers/EditSetRecordControllerSpec.scala:1325]: Best to remove the trailing space, or IntelliJ has trouble running the child scenarios separately.

File: test/controllers/EditSetRecordControllerSpec.scala [hunk 16]
```diff
@@ -1272,6 +1272,107 @@ class EditSetRecordControllerSpec extends BaseSpec {
...
+              valuesFromRecord(oci) ++ Map(
+                FieldNames.endDateDay -> "031"
+              )
+            val editRecordPageResponse = submitWhileLoggedIn("save", 1, oci, values)
+
+            status(editRecordPageResponse) mustBe SEE_OTHER
+            redirectLocation(editRecordPageResponse) mustBe Some(s"/edit-set/1/record/$oci/edit/save")
+
+            val getRecordResult = getRecordForEditingWhileLoggedIn(1, oci)
+            assertPageAsExpected(asDocument(getRecordResult), generateExpectedEditRecordPageFromRecord(oci))
```
Comment (tna-erasmos) [inline test/controllers/EditSetRecordControllerSpec.scala:1339]: I would have asserted the start date day more explicitly, as that is the focus of this scenario.

For example:

```
assertPageAsExpected(
  asDocument(getRecordResult),
  generateExpectedEditRecordPageFromRecord(oci)
    .copy(endDate = ExpectedDate("31", "12", "1962"))
)
```

File: test/controllers/EditSetRecordControllerSpec.scala [hunk 17]
```diff
@@ -1272,6 +1272,107 @@ class EditSetRecordControllerSpec extends BaseSpec {
...
+          }
+
+          "in month" in {
+
+            val oci = "COAL.2022.V1RJW.P"
+            val values =
+              valuesFromRecord(oci) ++ Map(
+                FieldNames.endDateMonth -> "012"
```
Comment (tna-erasmos) [inline test/controllers/EditSetRecordControllerSpec.scala:1347]: Maybe use a more realistic month, like '08'?

File: test/controllers/EditSetRecordControllerSpec.scala [hunk 18]
```diff
@@ -1272,6 +1272,107 @@ class EditSetRecordControllerSpec extends BaseSpec {
...
+            val getRecordResult = getRecordForEditingWhileLoggedIn(1, oci)
+
+            assertPageAsExpected(asDocument(getRecordResult), generateExpectedEditRecordPageFromRecord(oci))
+          }
+
+          "in year" in {
+
+            val oci = "COAL.2022.V1RJW.P"
+            val values =
+              valuesFromRecord(oci) ++ Map(
```
Comment (tna-erasmos) [inline test/controllers/EditSetRecordControllerSpec.scala:1363]: Maybe use a more realistic year, like '0940'? In that case, you'd have to ensure that the start date is before, of course, so perhaps use a start year of 940, as well?

File: test/controllers/EditSetRecordControllerSpec.scala [hunk 19]
```diff
@@ -1272,6 +1272,107 @@ class EditSetRecordControllerSpec extends BaseSpec {
...
+                FieldNames.endDateYear -> "01964"
+              )
+            val editRecordPageResponse = submitWhileLoggedIn("save", 1, oci, values)
+
+            status(editRecordPageResponse) mustBe SEE_OTHER
+            redirectLocation(editRecordPageResponse) mustBe Some(s"/edit-set/1/record/$oci/edit/save")
+
+            val getRecordResult = getRecordForEditingWhileLoggedIn(1, oci)
+            assertPageAsExpected(asDocument(getRecordResult), generateExpectedEditRecordPageFromRecord(oci))
```
Comment (tna-erasmos) [inline test/controllers/EditSetRecordControllerSpec.scala:1372]: I would have asserted the end date year more explicitly, as that is the focus of this scenario. If you went with a different start year, you'd need to change the expected value.

For example:

```
assertPageAsExpected(
  asDocument(getRecordResult),
  generateExpectedEditRecordPageFromRecord(oci)
    .copy(
      startDate = ExpectedDate("1", "1", "940"),
      endDate = ExpectedDate("31", "12", "940"))
)
```

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala [hunk 20]
```diff
@@ -83,6 +107,9 @@ class DateParserSpec extends WordSpec with MustMatchers {
         Seq("29/2/2022", "30/2/2022", "31/2/2022", "42/10/2022", "10/14/2022")
           .foreach(rawDate => parse(rawDate) mustBe empty)
       }
+      "given a negative year" in {
+        parse("1/11/-10") mustBe empty
+      }
     }
```
Comment (tna-erasmos) [inline test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala:113]: Should we have a scenario for when the year is zero?

**Review by tna-erasmos — CHANGES_REQUESTED** [review]

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala (hunk 20 above)
Comment (userjd) [reply test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala:113]: Done

File: test/controllers/EditSetRecordControllerSpec.scala (hunk 19 above)
Comment (userjd) [reply test/controllers/EditSetRecordControllerSpec.scala:1372]: Done

File: test/controllers/EditSetRecordControllerSpec.scala (hunk 18 above)
Comment (userjd) [reply test/controllers/EditSetRecordControllerSpec.scala:1363]: Done

File: test/controllers/EditSetRecordControllerSpec.scala (hunk 16 above)
Comment (userjd) [reply test/controllers/EditSetRecordControllerSpec.scala:1339]: Done

File: test/controllers/EditSetRecordControllerSpec.scala [hunk 21]
```diff
@@ -1272,6 +1272,107 @@ class EditSetRecordControllerSpec extends BaseSpec {
...
+              valuesFromRecord(oci) ++ Map(
+                FieldNames.startDateYear -> "01962"
+              )
+            val editRecordPageResponse = submitWhileLoggedIn("save", 1, oci, values)
+
+            status(editRecordPageResponse) mustBe SEE_OTHER
+            redirectLocation(editRecordPageResponse) mustBe Some(s"/edit-set/1/record/$oci/edit/save")
+
+            val getRecordResult = getRecordForEditingWhileLoggedIn(1, oci)
+            assertPageAsExpected(asDocument(getRecordResult), generateExpectedEditRecordPageFromRecord(oci))
```
[the remark this answers is not in the document]
Comment (userjd) [reply test/controllers/EditSetRecordControllerSpec.scala:1322]: Done

File: test/controllers/EditSetRecordControllerSpec.scala [hunk 22]
```diff
@@ -1272,6 +1272,107 @@ class EditSetRecordControllerSpec extends BaseSpec {
...
+              valuesFromRecord(oci) ++ Map(
+                FieldNames.startDateMonth -> "01"
+              )
+            val editRecordPageResponse = submitWhileLoggedIn("save", 1, oci, values)
+
+            status(editRecordPageResponse) mustBe SEE_OTHER
+            redirectLocation(editRecordPageResponse) mustBe Some(s"/edit-set/1/record/$oci/edit/save")
+
+            val getRecordResult = getRecordForEditingWhileLoggedIn(1, oci)
+            assertPageAsExpected(asDocument(getRecordResult), generateExpectedEditRecordPageFromRecord(oci))
```
[the remark this answers is not in the document]
Comment (userjd) [reply test/controllers/EditSetRecordControllerSpec.scala:1306]: Done

File: test/controllers/EditSetRecordControllerSpec.scala [hunk 23]
```diff
@@ -1272,6 +1272,107 @@ class EditSetRecordControllerSpec extends BaseSpec {
...
+            val oci = "COAL.2022.V1RJW.P"
+            val values =
+              valuesFromRecord(oci) ++ Map(
+                FieldNames.startDateDay -> "01"
+              )
+            val editRecordPageResponse = submitWhileLoggedIn("save", 1, oci, values)
+
+            status(editRecordPageResponse) mustBe SEE_OTHER
+            redirectLocation(editRecordPageResponse) mustBe Some(s"/edit-set/1/record/$oci/edit/save")
+            println(generateExpectedEditRecordPageFromRecord(oci))
```
[the remark this answers is not in the document]
Comment (userjd) [reply test/controllers/EditSetRecordControllerSpec.scala:1288]: Removed

File: test/controllers/EditSetRecordControllerSpec.scala [hunk 24]
```diff
@@ -1272,6 +1272,107 @@ class EditSetRecordControllerSpec extends BaseSpec {
           )
 
         }
+
+        "start date has a leading zero " when {
```
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/controllers/EditSetRecordControllerSpec.scala]: Confirmed.

File: test/controllers/EditSetRecordControllerSpec.scala (hunk 23 above)
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/controllers/EditSetRecordControllerSpec.scala]: Confirmed.

File: test/controllers/EditSetRecordControllerSpec.scala [hunk 25]
```diff
@@ -1272,6 +1272,107 @@ class EditSetRecordControllerSpec extends BaseSpec {
...
+            val getRecordResult = getRecordForEditingWhileLoggedIn(1, oci)
+            assertPageAsExpected(asDocument(getRecordResult), generateExpectedEditRecordPageFromRecord(oci))
+          }
+
+          "in year" in {
+
+            val oci = "COAL.2022.V1RJW.P"
+            val values =
+              valuesFromRecord(oci) ++ Map(
+                FieldNames.startDateYear -> "01962"
```
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/controllers/EditSetRecordControllerSpec.scala]: Confirmed.

File: test/controllers/EditSetRecordControllerSpec.scala (hunk 21 above)
[the remark this answers is not in the document]
Comment (tna-erasmos) [reply test/controllers/EditSetRecordControllerSpec.scala]: Confirmed.

File: test/controllers/EditSetRecordControllerSpec.scala (hunk 15 above)
Comment (tna-erasmos) [reply test/controllers/EditSetRecordControllerSpec.scala]: Confirmed.

File: test/controllers/EditSetRecordControllerSpec.scala (hunk 16 above)
Comment (tna-erasmos) [reply test/controllers/EditSetRecordControllerSpec.scala]: Confirmed.

File: test/controllers/EditSetRecordControllerSpec.scala (hunk 17 above)
Comment (tna-erasmos) [reply test/controllers/EditSetRecordControllerSpec.scala]: Confirmed.

File: test/controllers/EditSetRecordControllerSpec.scala (hunk 18 above)
Comment (tna-erasmos) [reply test/controllers/EditSetRecordControllerSpec.scala:1378]: Confirmed.

File: test/controllers/EditSetRecordControllerSpec.scala (hunk 19 above)
Comment (tna-erasmos) [reply test/controllers/EditSetRecordControllerSpec.scala]: Confirmed.

File: test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala (hunk 20 above)
Comment (tna-erasmos) [reply test/uk/gov/nationalarchives/omega/editorial/support/DateParserSpec.scala:116]: Confirmed.

**Review by tna-erasmos — APPROVED** [review]
LGTM

**Review by felixgb — APPROVED** [review]
